### ISSUE TITLE
How to filter on subscribers?

### ISSUE THREAD

**Issue by Robert-M-Muench** [ReactiveX/RxGo#270]
Observables can do many things on the items/messages they handle and then notify all subscribers.

However, how to do an Observable that will only notify a sub-set of all subscribers based on some condition(s)?

For example, how to notify only "2D" subscribers that were hit by a mouse-click?

**Comment by teivah** [comment]
Hey @Robert-M-Muench
That's the opposite approach. A subscriber can define a filter to define what is it interested in.
If you have multiple subscribers on the same observable, you also need to use a publish strategy to broadcast the elements.

As an example:

```go
type event struct {
	id        int
	eventType string
}

func main() {
	ch := make(chan rxgo.Item, 3)

	go func() {
		ch <- rxgo.Of(event{0, "3D"})
		ch <- rxgo.Of(event{1, "2D"})
		ch <- rxgo.Of(event{2, "2D"})
	}()

	observable := rxgo.FromChannel(ch, rxgo.WithPublishStrategy())

	subscriber2D := observable.Filter(func(i interface{}) bool {
		event := i.(event)
		return event.eventType == "2D"
	})
	subscriber2D.DoOnNext(func(i interface{}) {
		fmt.Printf("2D: %d\n", i.(event).id)
	})

	subscriber3D := observable.Filter(func(i interface{}) bool {
		event := i.(event)
		return event.eventType == "3D"
	})
	subscriber3D.DoOnNext(func(i interface{}) {
		fmt.Printf("3D: %d\n", i.(event).id)
	})

	observable.Connect(context.Background())
	select {}
}
```

Does it answer the question?

**Comment by Robert-M-Muench** [comment]
Hmm... not sure. To make it more concrete: Let's assume I have 100 subscribers in a mouse-click observable. And now only if a subscriber is hit, its callback should be called. How to do that?

**Comment by darrensapalo** [comment]
In your specific example, mouse clicks are [hot observables](https://github.com/darrensapalo/RxGo#hot-vs-cold-observables).

**Multicasting**

> Let's assume I have 100 subscribers in a mouse-click observable

The publish strategy teivah mentioned will indeed allow multiple subscribers to listen to a single observable. It is specified upon creation of the observable as demonstrated in his example.

> And now only if a subscriber is hit

You're supposed to write those conditions (i.e. "if a subscriber is hit") on the filter anonymous function, as @teivah described. 

Take note: the parameter for `filter` operator is solely its emission `i interface{}`. In teivah's example, it is an `event` struct. In your case, it is a mouse click event.

**Varying filter criteria**

If your subscribers have more factors than just the mouse click event, you'll need to combine your stream with the other factors (e.g. application state) so that it can be included in the stream of events / emission.

If the concern is how to scale up the code when you have 100 subscribers, a question is: do each of those subscribers have distinct filter conditions? You will need to define observable streams differentiated by their filter conditions- then you can reuse that observable for as many subscribers as you need.

**Comment by Robert-M-Muench** [comment]
Thanks for the example. Am I right, that in this case all 2D subscribers check the condition on every event? Let's assume only 1:100 is hit, which means 99% of the checks are unnecessary.

That's why I first want to filter and then just notify the hit subscribers.

**Comment by darrensapalo** [comment]
> in this case all 2D subscribers check the condition on every event?

That is correct. Both subscriptions (i.e. `subscriber2D` and `subscriber3D`) will run the filter operator per emission.

If you want to perform the filter only once rather than multiple times **and the filter condition is the same for two subscribers**, then you simply want to multicast.

That means that if you have 1:100 hit and you have let's say five subscribers, you can:

1. From your source observable, apply the desired filter operator
2. Multicast using the publish strategy only after the filter has been applied
3. Create your subscriptions (e.g. five subscribers)
4. Connect the observable after all subscriptions have been registered

This will cause the filter to only occur a total of 100 times, rather than 500 times.

**Comment by Robert-M-Muench** [comment]
I'm still confused. In my case step-1 changes on every mouse-move as I do a hit-check. So for every new mouse coordinate I check which subscribers were hit. I get back a list if hits.

So, what you suggest is, to do this hit-testing first. And from this result build an observable, where the hit observers are subscribed to, then emit the event and destroy everything for the next round?

I think what I want to do is: Super fast hit testing (already working), using a bitset for checking if a hit is subscribed would be simple too, and then just iterate over all hits and call the DoOnNext. This all looks to me much more like a specialized observable than anything else.

**Comment by Robert-M-Muench** [comment]
From http://reactivex.io/documentation/contract.html

Multiple Observers: If a second observer subscribes to an Observable that is already emitting items to a first observer, it is up to the Observable whether it will thenceforth emit the same items to each observer, or whether it will replay the complete sequence of items from the beginning to the second observer, or whether it will emit a wholly different sequence of items to the second observer. There is no general guarantee that two observers of the same Observable will see the same sequence of items.

The last sentence is exactly my case. And that decision is done by hit-testing. But as I understand RxGo, there is no official way to create an observable with a specific publishing strategy based on a hit-test.

**Comment by Robert-M-Muench** [comment]
https://github.com/ReactiveX/RxGo/blob/2c86562d9c8aba528e07af3fa1d00ec7ea4fc9e4/iterable_channel.go#L53-L77

I think that's the place where I want something different. Currently all subscribers are notified one-by-one. And I just want to notify a subset of the subscribers.

**Comment by darrensapalo** [comment]
I see, so your usage is you have mouse movement (a point/location) + some 2D object and you want to perform hit tests.

> But as I understand RxGo, there is no official way to create an observable with a specific publishing strategy based on a hit-test.

I think the reactive way of going about this is to transform your observable stream from `Observable<Point2D>` to `Observable<HitTest>`, wherein `HitTest` matches the following structure:

```go
type HitTest struct{
  mouseLocation Point2D
  hitbox Rectangle2D
}
```

And then using a `filter` prior to subscription.

```go
oneSubscriber := observable.Filter(func(i interface{}) bool {
	event := i.(event)
	return event.hitbox.Contains(event.mouseLocation)
})
```

If the objects move, you might want to  consider `combineLatest` since you'll need the latest hitbox and the latest mouse movement.

## Opinion

In my personal experience, I have rarely found that the toolset/language provided by the Rx paradigm to be lacking (i.e. I don't think that the reactivex community overlooked a use-case for a publishing strategy that has filtering baked into it). In my opinion, it is more likely that I am still grappling with reactivex' paradigm + language than the community overlooking some general use-case*.

Of course you are allowed to make your own operators, publishing strategies, observable constructors, etc. though I would prefer to use the already available operators and observables. This allows me to leverage unit-tested code in the existing toolset vs. me developing my own operators/publishing strategies and plan out/address concurrency issues as well.

_*I also understand RxGo is fairly new and in-progress with limited tests and code, but I am speaking about the completeness of the reactivex framework as a language, not as an implementation_

## How I would deal with your situation

Your problem definition:

1. 1 observable source (mouse movements)
2. multiple subscribers (2d rectangular objects)
3. constraint: subscription should only run if hitbox check passes.


From my experience, I would answer (1) with multicasting i.e. publish strategy as demonstrated by teivah above, answer (2) and (3) with `filter` operator. 

In your use case, you need the location AND the hitbox rectangle-- you cannot perform hitbox testing with only the mouse location. This is what I meant by:

> Varying filter criteria
> 
> If your subscribers have more factors than just the mouse click event, you'll need to combine your stream with the other factors (e.g. application state) so that it can be included in the stream of events / emission.

If you do decide to design a publishing strategy that supports varying filtering criteria for varying subscribers, I am curious to see how that would be designed.

Potential scenarios are:

* Number of filtering criteria = number of subscribers, with no filtering overlap (each have their own filtering cases)
* Number of filtering criteria < number of subscribers, with some filtering overlap

**Comment by Robert-M-Muench** [comment]
Thx, for you long answer.

The problem with using any Filter is that it acts on every subscriber.

Let's assume I have 5120 x 2880 = 14'745'600 pixel and for each pixel on the subscriber. So, for every mouse-location, only one subscriber is hit at the same time.

The problem is not to model the hittest, that's easy and super fast, even with 14 Mio. rectangles/points.

The problem is, that 14'745'600 subscribers will perform the hittest, whereas I can get the hit subscribers up-front with on simple query. And know this list, I know exactly which subscribers I just need to trigger, not caring about all the rest at all.

In pseudo-code:
```
 subscribersHit = query(x,y)
 for x = 0; x < len(subscribersHit); x++ {
   subscribersHit[x].DoOnNext()
 } 
```

Using Filter, all subscribers are informed about (x,y) and check if they are hit. Very inefficient.

It seems that most RX implementations assume that all subscribers are informed. There is no way to provide something like a querySubscribersInScope function, which returns a list of subscribers to inform.

And, yes, I would like to use that standard stuff too, and not fiddle around with my own code.

**Comment by darrensapalo** [comment]
Thank you as well for providing more context.

> Let's assume I have 5120 x 2880 = 14'745'600 pixel and for each pixel on the subscriber. So, for every mouse-location, only one subscriber is hit at the same time.

In the specific case of having subscriptions per pixel vs. subscriptions per entity that has a hitbox, I can see why you would need a custom publishing strategy.

I'm curious, is there a specific reason why you need a subscription per pixel rather than per entity to hit test?

Maintaining 15M subscriptions would also take up so much memory space rather than having N subscriptions where N is the number of entities to hittest. 

My general assumption (which may be invalid in your use case) is that N would often be lower than the image screen size, but cases may vary (e.g. You're designing a game engine that will regularly perform these operations- or you're designing some kind of particle/image processing library).

**Comment by Robert-M-Muench** [comment]
I just drove the example to an extreme, to make clear, that in my use-case the number of subscribers vs. the number of affected subscribers is 100:1 or 1000:1. Hence, I don't want to spend most of the run-time on subscribers finding out, that they are not affected.

I think all GUI related things fall into this category. You have hundreds of GUI elements, but the user only interacts with a fraction at one time. I think this is a pretty common pattern, that's why I'm a bit confused, it's not covered.

The RX community seems to see this use-case (see doc link I posted) but a user-defined publishing-strategy doesn't seem to be implemented. Or I didn't find out how to do it. Any idea?

**Comment by darrensapalo** [comment]
Haven't done something like that for RxGo before, but working on top of your pseudo-code and fleshing out the dependencies, I find that you'll need to make the following revisions:

```
 subscribersHit = query(subscribers, mapping, x, y)
 for x = 0; x < len(subscribersHit); x++ {
   subscribersHit[x].DoOnNext()
 } 
```

where `mapping` is a mapping from a subscriber to a hitbox.

## Dependencies

In your query, you'll likely need a list of `subscribers` as I assume the filtering conditions (hitbox + x,y) for the subscribers vary. 

You'll also need a `mapping` parameter because `subscribers` are highly generic-- they're just a channel that emits items. It doesn't know exactly which subscriber owns which hitbox.

## Observable design

Based on my understanding the design of observables by reactivex, at the start of defining a source observable, **an observable does not know who the subscribers will be**. This abstraction is intended so that your observable has functional purity/idempotence. It doesn't have some kind of external or global state dependency when processing it and the behavior will be the same for the same input.

This constraint is broken because of the new dependency (i.e. `mapping`) that changes the behavior of the observable. Used correctly, it isn't exactly a huge behavior change as we know it is an optimization. But used incorrectly (e.g. not registering the mapping for the subscribers) and the observable suddenly will not work.

## Additional subscription step

The challenging part here is generating the `mapping`- it is easier said than done. 

By making `mapping` a parameter for the publishing strategy, you'll find yourself trying to figure out how to add an additional registration process for the subscriber. You'll then be introducing something on-top of the agreed upon contract of how observables + subscribers work:

```
mapping := makeEmptyMapping()
yourSourceObservable := makeObservableWithCustomPublishingStrategy(mapping)

someSubscriber := func(i interface{}) {
    fmt.Printf("mouseclick: %d\n", i.(event).id)
}

yourSourceObservable.DoOnNext(someSubscriber)
registerMappingWithHitbox(mapping, &someSubscriber)
```

The standard subscription process for observables only requires lambda function `someSubscriber` as a parameter. In the above example for `DoOnNext`, it does not accept any other parameter. The missing step that you would need is pairing that subscription of that specific subscriber to own a specific hitbox.

The above pseudocode is wrong at the last two lines, because the subscription registration by the `DoOnNext` doesn't use addresses. If it did use addresses, it would be simpler to map the subscriber to which hitbox it is related to. But that would be changing how observables and subscriptions work altogether.

Not sure how you'll move forward after this, but I do hope that gives some insight on to your work. Good luck! 👍

**Comment by Robert-M-Muench** [comment]
We don't have to care about the hit-testing. It's done differently is not part of the RX handling. Every subscriber carries its own hit-box, which is registered in a rTree, which then is queried. So, you feed in an (X, Y) point and get back the hit subscribers. If a hit-box changes, this is reflected in the rTree.

>  an observable does not know who the subscribers will be

Yes, that's right. In that context, we can see my use-case, like not having an explicit subscription action from every subscriber, which then is static until changed by every subscriber again. Instead, we have a subscriber-function (here the query), yielding a different & dynamic set of subscribers to act on. If a subscriber is part of the query, it depends on other factors than the .subscribe action. It can depend on what-ever makes sense. Such a subscriber-function can be seen as a lambda.

The subscriber-function is like a proxy between the observable and the subscribers, doing whatever makes sense.

**Comment by Robert-M-Muench** [comment]
I solved this now in a different way because RX is really not prepared to filter the subscriber list inside the observable. I would have to create a new observable which has a huge interface, while I just need a simple change at one point.

I have now one subscriber to the stream, that does the filtering, and just calls a callback function of all hit objects. Using Go's interfaces, a dynamic check if the interface is supported is used. This makes it easy for a user, who just needs to support the interface to get notified.
